### PR TITLE
Improve crash handling

### PR DESCRIPTION
**Pull request by huffman**
The goal here was to increase the crash coverage of BugSplat. More work could be done here, but this is a good start.

### CODE REVIEW

File: interface/src/CrashReporter.cpp [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    signal(SIGSEGV, handleSignal);
+    signal(SIGABRT, handleSignal);
+    _set_purecall_handler(handlePureVirtualCall);
+    _set_invalid_parameter_handler(handleInvalidParameter);
+    _set_new_mode(1);
+    _set_new_handler(handleNewError);
+
+    // Disable WER popup
+    //SetErrorMode(SEM_FAILCRITICALERRORS | SEM_NOGPFAULTERRORBOX | SEM_NOOPENFILEERRORBOX);
+    //_set_abort_behavior(0, _WRITE_ABORT_MSG | _CALL_REPORTFAULT);
```
Comment (huffman) [inline interface/src/CrashReporter.cpp]: Leaving this here for now. If anyone has thoughts on whether we should leave the WER handling enabled I'm open ears. The WER dialog is the "application has stopped responding..." popup on Windows.

**Comment by huffman** [comment]
Working on testing details.

**Comment by hifi-gustavo** [comment]
Build console output for commit: afe502e1d22d56842906df11612ab852f4bb38a6 available at: https://gist.github.com/7c1034e5cfde3139ee9fa1faa29b3564<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7639.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7639.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2dfe96a654761fc474c7a2ce0aab31e25c1faef7 available at: https://gist.github.com/7c1034e5cfde3139ee9fa1faa29b3564<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7639.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7639.dmg

**Comment by ZappoMan** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2dfe96a654761fc474c7a2ce0aab31e25c1faef7 available at: https://gist.github.com/7c1034e5cfde3139ee9fa1faa29b3564<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7639.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7639.dmg

**Comment by ZappoMan** [comment]
This comment was originally by @huffman but I wanted to remove it from the PR description since we don't want it to be part of the release notes.
---------
Relevant comment from the code about patching SetUnhandledExceptionFilter:

> SetUnhandledExceptionFilter can be overridden by the CRT at the point that an error occurs. More information can be found here: http://www.codeproject.com/Articles/154686/SetUnhandledExceptionFilter-and-the-C-C-Runtime-Li 
A fairly common approach is to patch the SetUnhandledExceptionFilter so that it cannot be overridden and so that the applicaiton can handle it itself.
The CAPIHook class referenced in the above article is not openly available, but a similar implementation can be found here: http://blog.kalmbach-software.de/2008/04/02/unhandled-exceptions-in-vc8-and-above-for-x86-and-x64/

File: interface/src/CrashReporter.cpp [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+}
+
+LPTOP_LEVEL_EXCEPTION_FILTER WINAPI
+noop_SetUnhandledExceptionFilter(
+    LPTOP_LEVEL_EXCEPTION_FILTER lpTopLevelExceptionFilter)
+{
+    return nullptr;
+}
+
+_purecall_handler __cdecl
```
Comment (ZappoMan) [inline interface/src/CrashReporter.cpp]: coding standard brace placement:

```
_purecall_handler __cdecl noop_set_purecall_handler(_purecall_handler pNew) {
...
```

File: interface/src/CrashReporter.cpp [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    // Throw so BugSplat can handle
+    throw("ERROR: Invalid parameter");
+}
+
+int handleNewError(size_t size) {
+    // Throw so BugSplat can handle
+    throw("ERROR: Errors calling new");
+}
+
+LPTOP_LEVEL_EXCEPTION_FILTER WINAPI
```
Comment (ZappoMan) [inline interface/src/CrashReporter.cpp]: coding standard brace placement:

```
LPTOP_LEVEL_EXCEPTION_FILTER WINAPI noop_SetUnhandledExceptionFilter(LPTOP_LEVEL_EXCEPTION_FILTER lpTopLevelExceptionFilter) {
```

File: interface/src/CrashReporter.cpp (hunk 1 above)
Comment (ZappoMan) [inline interface/src/CrashReporter.cpp]: If you're saying we could trap that event and send ourselves a bugsplat, then yes, I think we should trap that event.

What's the downside?

**Comment by ZappoMan** [comment]
Are you going to add some "Developer" menus to force these various crashes? is that the intent of `CrashHelpers.h`? That seems like a reasonable thing to do to test these new crash modes.

**Comment by kevinmthomas-carpool** [comment]
## QA Test
**PR Build:** 7639
**Testing Architecture:** 64-bit PC Desktop
**Testing OS:** Windows 10
**Testing Locations:** Sandbox, Playa, Cellscience, Entry
**Smoke Test Results:** Successful test.

                                  |Playa    |Cellscience |Entry  
----------------------------------|---------|------------|--------
Render Rate:                      |63       |64          |61           
Present Rate:                     |60       |60          |60          
Simrate:                          |62       |63          |61           
Render Stats Num Buffers:         |3051     |2118        |7596              
Render Stats gpu::Buffer Memory:  |58.3Mb   |22.8Mb      |49.4Mb              
Render Stats Num Textures:        |273      |164         |390              
Render Stats gpu::Texture Memory: |853.1Mb  |267.7Mb     |540.5Mb              
Render Stats Triangles:           |325k     |370k        |298k            
Render Stats Drawcalls:           |23532    |14090       |4006             
Render Stats Items:               |134      |76          |120  
GPU-Z Memory Size                 |4096Mb   |4096Mb      |4096Mb
GPU-Z Initial Memory		  |237Mb    |237Mb	 |237Mb
GPU-Z Memory Used		  |2459Mb   |1521Mb      |1965Mb
GPU-Z Memory Load		  |11%      |8%          |7%

**sandbox ‘createtestblocks.js’ Result:** Successful test.
**sandbox Import FBX Result:** Successful test.
**sandbox ‘harmonicOscillator.js’ Result:** Successful test.
**sandbox Echo Server Audio Result:** Successful test.
**sandbox ‘reverbTest.js’ Result:** Successful test.
**Additional Test Results Requested:** Unable to cause any user crashes as the build was very stable.
**Successful Logout/Crash:** Successful logout.

File: interface/src/CrashReporter.cpp (hunk 1 above)
Comment (huffman) [inline interface/src/CrashReporter.cpp]: I have observed 2 cases where WER pops up:

1. WER pops up, but if you press cancel BugSplat will display
2. WER pops up, buf if you press cancel BugSplat does not display

WER gives an opportunity to debug, so there is some benefit for scenario #2. That said, you should still be able to get a minidump from `~/AppData/Local/Crash Dumps` if you want to look at the crash (although I haven't checked this). If we are okay with that then I think disabling the WER popup makes sense.

**Comment by huffman** [comment]
@ZappoMan Yes, I was using those manually but am going to add them to the dev menu for testing.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 88150f9c82b7c0e296be5d9cd99888b15f3485ab available at: https://gist.github.com/7c1034e5cfde3139ee9fa1faa29b3564<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7639.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7639.dmg

**Comment by huffman** [comment]
Testing for this is blocked by https://github.com/highfidelity/hifi/pull/7652/files - once that is built and merged I'll have this PR rebuilt so it includes the new menus

**Comment by huffman** [comment]
Results using the new menu items in https://github.com/highfidelity/hifi/pull/7652#issuecomment-209540776: 

(X means the crash was caught by BugSplat)
(threaded means the crash did not occur on the main thread)

| Crash Type               | Old Build  | This Build |
|--------------------------|------------|------------|
| Pure Virtual             |     | X     |
| Pure Virtual (threaded)  |     |     |
| Double Free              |     |     |
| Double Free (threaded)   |     | X     |
| Abort                    |     | X     |
| Abort (threaded)         |     |     |
| Null Deref               | X     | X     |
| Null Deref (threaded)    |     | X     |
| Out of Bounds            | X     | X     |
| Out of Bounds (threaded) |     | X     |
| New Fault                | X     | X     |
| New Fault (threaded)     |     |     |

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2d39551a35570886a6a49385dcc1341de8aa3fc1 available at: https://gist.github.com/7c1034e5cfde3139ee9fa1faa29b3564<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7639.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7639.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: e76ed6f73a7541adc0ae3c5458ea7e01c0e0ec37 available at: https://gist.github.com/7c1034e5cfde3139ee9fa1faa29b3564<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7639.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7639.dmg

**Comment by OmegaHeron** [comment]
Testing Windows 7 (Specifically testing new crash options + bugsplat coverage)

Deadlock - Bugsplat presented.
Pure Virtual Function Call - Bugsplat presented.
Pure Virtual Function Call Threaded - No bugsplat - windows dialog to close only.
Double Free - No bugsplat - windows dialog to close only.
Double Free Threaded - No bugsplat - windows dialog to close only.
Abort - Bugsplat presented.
Abort Threaded - Bugsplat presented.
Null Dereference - Bugsplat presented.
Null Dereference Threaded - No bugsplat - windows dialog to close only.
Out of Bounds Vector Access - Bugsplat presented.
Out of Bounds Vector Access Threaded - No bugsplat - windows dialog to close only.
New Fault - Bugsplat presented.
New Fault Threaded - Windows dialog + windows send crash report + bugsplat all presented.

**Comment by OmegaHeron** [comment]
Additional background - I do not have MSVC/debug tools installed on this machine - I know from past experience where I did have those tools installed you get somewhat differing behavior/dialogs.  But - this is a more end user setup vs devel since I want to test from "normal user" mode and not have devel tools hide things or create things ;)

**Comment by huffman** [comment]
Ok, so it looks like the coverage is better, but there is a discrepancy between our results (as in the other PR). I will continue to see where we can improve coverage and consistency in catching these crashes.

**Pull request merged**